File: airborne_sdk_iOS/hyper-ota/Airborne/AirborneObjC/FileUtil/AJPFileUtil.m
//
//  AJPFileUtil.m
//  Airborne
//
//  Created by Balaganesh S on 23/09/25.
//

#import "AJPFileUtil.h"

@interface AJPFileUtil()

@property NSString *workspace;
@property NSBundle *parentBundle;

@end


@implementation AJPFileUtil

- (instancetype)initWithWorkspace:(NSString *)workspace baseBundle:(NSBundle * _Nullable)baseBundle {
    self = [super init];
    if (self) {
        self.workspace = workspace;
        self.parentBundle = baseBundle;
        NSString* pathForClientBundle = [baseBundle pathForResource:workspace ofType:@"bundle"];
        if (pathForClientBundle) {
            _assetsBundle = [NSBundle bundleWithPath:pathForClientBundle];
        } else {
            _assetsBundle = baseBundle;
        }
    }
    return self;
}

#pragma mark - File Path

- (NSString *)filePathInBundleForFileName:(NSString *)fileName {

    NSArray *fileNameComponents = [fileName componentsSeparatedByString:@"."];

    NSString *fileNameString;

    for (int i=0; i<fileNameComponents.count-1; i++) {
        if (i==0) {
            fileNameString = [NSString stringWithFormat:@"%@",fileNameComponents[i]];
        } else {
            fileNameString = [NSString stringWithFormat:@"%@.%@",fileNameString,fileNameComponents[i]];
        }
    }
    
    NSString *filePathInBundle = [self.assetsBundle pathForResource:fileNameString ofType:fileNameComponents.lastObject];
    if (filePathInBundle) {
        return filePathInBundle;
    }

    filePathInBundle = [self.parentBundle pathForResource:fileNameString ofType:fileNameComponents.lastObject];
    if (filePathInBundle) {
        return filePathInBundle;
    }

    return [[NSBundle mainBundle] pathForResource:fileNameString ofType:fileNameComponents.lastObject];;
}

- (NSString *)fullPathInStorageForFilePath:(NSString*)filePath inFolder:(NSString*)folderName {

    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSLibraryDirectory, NSUserDomainMask, YES);
    NSString* rootPath = paths[0];

    NSString* path = rootPath;

    //This just in case their is folder along with the file name.
    if (folderName.length > 0) {
        rootPath = [rootPath stringByAppendingPathComponent:folderName];
    }
    rootPath = [rootPath stringByAppendingPathComponent:self.workspace];
    
    NSArray *subPaths = [filePath componentsSeparatedByString:@"/"];
    if (subPaths.count > 1) {
        for (NSInteger i = 0; i < subPaths.count - 1; i++) {
            rootPath = [rootPath stringByAppendingPathComponent:subPaths[i]];
        }
    }
    
    [self createFolderIfDoesNotExist:rootPath];

    path = [rootPath stringByAppendingPathComponent:[filePath lastPathComponent]];

    return path;
}

#pragma mark - File Read

- (NSString *)loadFile:(NSString *)filePath folder:(NSString *)folder withLocalAssets:(Boolean)local error:(NSError **)error {

    NSError *fileReadError = nil;
    NSData *fileData = [self getFileDataForFileName:filePath inFolder:folder withAssetsFromLocal:local error:&fileReadError];

    if (!fileData) {
        if (error) {
            *error = fileReadError ?: [NSError errorWithDomain:@"in.juspay.Airborne" code:1001 userInfo:@{NSLocalizedDescriptionKey: @"Failed to read file content."}];
        }
        return @"";
    }
    
    NSString *fileDataString = [[NSString alloc] initWithData:fileData encoding:NSUTF8StringEncoding];
    
    if (!fileDataString) {
        if (error) {
            *error = [NSError errorWithDomain:@"in.juspay.Airborne" code:1002 userInfo:@{NSLocalizedDescriptionKey: @"Failed to convert file data to string."}];
        }
        return @"";
    }

    return fileDataString;
}

- (NSData *)getFileDataForFileName:(NSString *)fileName inFolder:(NSString *)folderName withAssetsFromLocal:(Boolean)local error:(NSError*__autoreleasing*) error {
    NSData *data = nil;

    if (!local) {
        data = [self getFileDataFromInternalStorage:fileName inFolder:folderName error:error];
    }
    if (data == nil) {
        data = [self getFileDataFromBundle:[fileName lastPathComponent] error:error];
    }
    return data;
}

- (NSData *)getFileDataFromBundle:(NSString *)fileName error:(NSError * __autoreleasing *)error {

    NSData *fileContents = nil;
    NSFileManager* manager = [NSFileManager defaultManager];

    NSString *filePathInBundle = [self filePathInBundleForFileName:fileName] ;
    if ([manager fileExistsAtPath:filePathInBundle]) {
        //Reading the contents of file stored in bundle.
        fileContents = [NSData dataWithContentsOfFile:filePathInBundle options:NSDataReadingMappedIfSafe error:error];
    }

    return fileContents;
}

- (NSData *)getFileDataFromInternalStorage:(NSString *)fileName inFolder:(NSString *)folderName error:(NSError**) error {
    NSData *fileContents = nil;
    NSFileManager* manager = [NSFileManager defaultManager];
    NSString *filePath = [self fullPathInStorageForFilePath:fileName inFolder:folderName];
    if ([manager fileExistsAtPath:filePath]) {
        fileContents = [NSData dataWithContentsOfFile:filePath options:NSDataReadingMappedIfSafe error:error];
    }

    return fileContents;
}

- (id _Nullable)getDecodedInstanceForClass:(Class)className withContentOfFileName:(NSString *)fileName inFolder:(NSString *)folderName error:(NSError *__autoreleasing *)error {
    NSData *fileData = [self getFileDataFromInternalStorage:fileName inFolder:folderName error:error];
    if (fileData != nil) {
        id decodedObject = [NSKeyedUnarchiver unarchivedObjectOfClass:className fromData:fileData error:error];
        return decodedObject;
    }

    return nil;
}

#pragma mark - File Write

- (BOOL)writeInstance:(id<NSCoding>)object fileName:(NSString *)fileName inFolder:(NSString *)folderName error:(NSError * __autoreleasing *) error {
    NSData *encodedData = [NSKeyedArchiver archivedDataWithRootObject:object requiringSecureCoding:YES error:error];
    if (!encodedData) {
        NSLog(@"Failed to encode object");
        return NO;
    }
    return [self saveFileWithData:encodedData fileName:fileName folderName:folderName error:error];
}

- (BOOL)saveFileWithData:(NSData *)content fileName:(NSString *)fileName folderName:(NSString *)folderName error:(NSError**) error {
    NSFileManager* manager = [NSFileManager defaultManager];

    NSString *filePath = [self fullPathInStorageForFilePath:fileName inFolder:folderName];

    if ([manager fileExistsAtPath:filePath]) {
        [manager removeItemAtPath:filePath error:error];
    }
    
    return [content writeToFile:filePath options:NSDataWritingAtomic error:error];

}

- (BOOL)moveFileToInternalStorage:(NSURL *)source fileName:(NSString *)fileName folderName:(NSString *)folderName error:(NSError * __autoreleasing *)error {
    NSFileManager* manager = [NSFileManager defaultManager];

    NSString *filePath = [self fullPathInStorageForFilePath:fileName inFolder:folderName];
    NSURL* destinationURL = [NSURL URLWithString:filePath];

    if ([manager fileExistsAtPath:filePath]) {
        return [manager replaceItemAtURL:destinationURL withItemAtURL:source backupItemName:nil options:NSFileManagerItemReplacementUsingNewMetadataOnly resultingItemURL:nil error:error];

    }
    return [manager moveItemAtURL:source toURL:destinationURL error:error];
}

- (void)createFolderIfDoesNotExist:(NSString *)folderName {
    if (folderName.length > 0) {
        if (![[NSFileManager defaultManager] fileExistsAtPath:folderName]) {
            [[NSFileManager defaultManager] createDirectoryAtPath:folderName
                                      withIntermediateDirectories:YES
                                                       attributes:nil
                                                            error:nil];
        }
    }
}

- (BOOL)deleteFile:(NSString *)fileName inFolder:(NSString *)folder error:(NSError * __autoreleasing *)error {
    NSString* path = [self fullPathInStorageForFilePath:fileName inFolder:folder];
    NSFileManager* fileManager = [NSFileManager defaultManager];
    return [fileManager removeItemAtPath:path error:error];
}

@end

File: airborne_sdk_iOS/hyper-ota/Airborne/AirborneSwiftCore/AJPFileUtil.swift
//
//  AJPFileUtil.swift
//  Airborne
//

import Foundation

/// A utility class for local file IO operations.
/// Handles the core workspace abstractions, directory management, and generic decoding.
@objc open class AJPFileUtil: NSObject {
    
    private let workspace: String
    private let parentBundle: Bundle?
    
    @objc public let assetsBundle: Bundle?
    
    /// Initializes a new instance of AJPFileUtil.
    /// - Parameters:
    ///   - workspace: The name of the workspace folder.
    ///   - baseBundle: The parent bundle. If nil, it defaults to the main app bundle.
    @objc public init(workspace: String, baseBundle: Bundle?) {
        self.workspace = workspace
        self.parentBundle = baseBundle
        
        if let baseBundle = baseBundle,
           let pathForClientBundle = baseBundle.path(forResource: workspace, ofType: "bundle"),
           let embeddedBundle = Bundle(path: pathForClientBundle) {
            self.assetsBundle = embeddedBundle
        } else {
            self.assetsBundle = baseBundle
        }
        super.init()
    }
    
    // MARK: - File Path
    
    /// Returns the absolute bundle path for a file if it exists.
    /// It systematically checks the asset bundle, then the parent bundle, and finally the main app bundle.
    /// - Parameter fileName: The name of the file (including varying extensions) to search for.
    /// - Returns: The absolute path string within the Bundle, or nil if the file is not found.
    @objc public func filePathInBundleForFileName(_ fileName: String) -> String? {
        let fileNameComponents = fileName.components(separatedBy: ".")
        
        guard !fileNameComponents.isEmpty else { return nil }
        
        var fileNameString = ""
        for i in 0..<(fileNameComponents.count - 1) {
            if i == 0 {
                fileNameString = "\(fileNameComponents[i])"
            } else {
                fileNameString = "\(fileNameString).\(fileNameComponents[i])"
            }
        }
        
        let type = fileNameComponents.last
        
        if let path = assetsBundle?.path(forResource: fileNameString, ofType: type) {
            return path
        }
        
        if let path = parentBundle?.path(forResource: fileNameString, ofType: type) {
            return path
        }
        
        return Bundle.main.path(forResource: fileNameString, ofType: type)
    }
    
    /// Generates the absolute file path within the application's internal Sandbox library.
    /// - Parameters:
    ///   - filePath: The relative path or name of the file.
    ///   - folderName: An optional subfolder name to isolate the file.
    /// - Returns: The fully constructed absolute file path.
    @objc public func fullPathInStorageForFilePath(_ filePath: String, inFolder folderName: String?) -> String {
        let paths = NSSearchPathForDirectoriesInDomains(.libraryDirectory, .userDomainMask, true)
        guard var rootPath = paths.first else { return "" }
        
        if let folderName = folderName, !folderName.isEmpty {
            rootPath = (rootPath as NSString).appendingPathComponent(folderName)
        }
        
        rootPath = (rootPath as NSString).appendingPathComponent(self.workspace)
        
        let subPaths = filePath.components(separatedBy: "/")
        if subPaths.count > 1 {
            for i in 0..<(subPaths.count - 1) {
                rootPath = (rootPath as NSString).appendingPathComponent(subPaths[i])
            }
        }
        
        createFolderIfDoesNotExist(rootPath)
        
        return (rootPath as NSString).appendingPathComponent((filePath as NSString).lastPathComponent)
    }
    
    // MARK: - File Read
    
    /// Loads the text content of a file as a String.
    /// It searches either the internal storage or the application's bundle depending on the parameters.
    /// - Parameters:
    ///   - filePath: The name of the file to load.
    ///   - folder: The subfolder containing the file.
    ///   - local: If true, it skips internal storage and only checks the bundle.
    /// - Throws: An `NSError` if the file cannot be found or if its content isn't valid UTF-8.
    /// - Returns: The file's content as a String.
    @objc open func loadFile(_ filePath: String, folder: String, withLocalAssets local: Bool) throws -> String {
        var fileReadError: NSError? = nil
        let fileData: Data?
        
        do {
            fileData = try getFileDataForFileName(filePath, inFolder: folder, withAssetsFromLocal: local)
        } catch {
            fileData = nil
            fileReadError = error as NSError
        }
        
        guard let data = fileData else {
            throw fileReadError ?? NSError(domain: "in.juspay.Airborne", code: 1001, userInfo: [NSLocalizedDescriptionKey: "Failed to read file content."])
        }
        
        guard let fileDataString = String(data: data, encoding: .utf8) else {
            throw NSError(domain: "in.juspay.Airborne", code: 1002, userInfo: [NSLocalizedDescriptionKey: "Failed to convert file data to string."])
        }
        
        return fileDataString
    }
    
    /// Retrieves the raw Data of a file, checking internal storage first and falling back to the bundle.
    /// - Parameters:
    ///   - fileName: The name of the file to retrieve.
    ///   - folderName: The subfolder containing the file.
    ///   - local: If true, it skips internal storage and only checks the bundle.
    /// - Throws: An `NSError` if the file does not exist.
    /// - Returns: The file's raw Data.
    @objc public func getFileDataForFileName(_ fileName: String, inFolder folderName: String?, withAssetsFromLocal local: Bool) throws -> Data {
        if !local {
            if let data = try? getFileDataFromInternalStorage(fileName, inFolder: folderName) {
                return data
            }
        }
        return try getFileDataFromBundle((fileName as NSString).lastPathComponent)
    }
    
    /// Retrieves the raw Data of a file localized exclusively within the application's bundle.
    /// - Parameter fileName: The name of the file to retrieve.
    /// - Throws: An `NSError` if the file does not exist in the bundle.
    /// - Returns: The file's raw Data securely mapped into memory.
    @objc public func getFileDataFromBundle(_ fileName: String) throws -> Data {
        let fileManager = FileManager.default
        if let filePathInBundle = filePathInBundleForFileName(fileName), fileManager.fileExists(atPath: filePathInBundle) {
            return try Data(contentsOf: URL(fileURLWithPath: filePathInBundle), options: .mappedIfSafe)
        }
        throw NSError(domain: NSCocoaErrorDomain, code: NSFileReadNoSuchFileError, userInfo: nil)
    }
    
    /// Retrieves the raw Data of a file located solely in the application's internal storage sandbox.
    /// - Parameters:
    ///   - fileName: The name of the file to retrieve.
    ///   - folderName: The subfolder where the file is stored.
    /// - Throws: An `NSError` if the file does not exist.
    /// - Returns: The file's raw Data securely mapped into memory.
    @objc public func getFileDataFromInternalStorage(_ fileName: String, inFolder folderName: String?) throws -> Data {
        let fileManager = FileManager.default
        let filePath = fullPathInStorageForFilePath(fileName, inFolder: folderName)
        if fileManager.fileExists(atPath: filePath) {
            return try Data(contentsOf: URL(fileURLWithPath: filePath), options: .mappedIfSafe)
        }
        throw NSError(domain: NSCocoaErrorDomain, code: NSFileReadNoSuchFileError, userInfo: nil)
    }
    
    /// Reads and decodes an archived object from a file in the application's internal storage.
    /// - Parameters:
    ///   - className: The class type of the object you expect to decode (e.g., `MyModel.self`).
    ///   - fileName: The name of the file containing the archived data.
    ///   - folderName: The folder where the file is located within the workspace.
    /// - Throws: An `NSError` if the file cannot be read or if decoding fails.
    /// - Returns: The fully decoded object. You will typically cast it to your expected type after receiving it.
    @objc public func getDecodedInstanceForClass(_ className: AnyClass, withContentOfFileName fileName: String, inFolder folderName: String) throws -> Any {
        let fileData = try getFileDataFromInternalStorage(fileName, inFolder: folderName)
        guard let decoded = try NSKeyedUnarchiver.unarchivedObject(ofClasses: [className], from: fileData) else {
            throw NSError(domain: "in.juspay.Airborne", code: 1003, userInfo: [NSLocalizedDescriptionKey: "Failed to decode object or object was nil"])
        }
        return decoded
    }
    
    // MARK: - File Write
    
    /// Encodes an object that conforms to `NSCoding` and securely saves it to internal storage.
    /// - Parameters:
    ///   - object: The object to serialize and save.
    ///   - fileName: The target file name for the saved data.
    ///   - folderName: The subfolder to store the file in.
    /// - Throws: An `NSError` if the object fails to encode or if the file cannot be written.
    @objc public func writeInstance(_ object: NSSecureCoding, fileName: String, inFolder folderName: String?) throws {
        // Modern secure encoding
        let data = try NSKeyedArchiver.archivedData(withRootObject: object, requiringSecureCoding: true)
        try saveFileWithData(data, fileName: fileName, folderName: folderName)
    }
    
    /// Writes raw Data directly to a file in internal storage, atomically replacing any existing file.
    /// - Parameters:
    ///   - content: The raw Data to write.
    ///   - fileName: The target file name.
    ///   - folderName: The subfolder to store the file in.
    /// - Throws: An `NSError` if the file cannot be written.
    @objc public func saveFileWithData(_ content: Data, fileName: String, folderName: String?) throws {
        let filePath = fullPathInStorageForFilePath(fileName, inFolder: folderName)
        try content.write(to: URL(fileURLWithPath: filePath), options: .atomic)
    }
    
    /// Safely moves a file from a given source URL to the application's internal storage sandbox.
    /// If a file already exists at the destination, it is atomically replaced.
    /// - Parameters:
    ///   - source: The absolute URL of the external or temporary file to move.
    ///   - fileName: The intended file name (including extension) at the destination.
    ///   - folderName: An optional subfolder namespace within the workspace to isolate the file.
    /// - Throws: An `NSError` if the file manager fails to move or replace the item.
    @objc public func moveFileToInternalStorage(_ source: URL, fileName: String, folderName: String?) throws {
        let fileManager = FileManager.default
        let filePath = fullPathInStorageForFilePath(fileName, inFolder: folderName)
        let destinationURL = URL(fileURLWithPath: filePath)
        
        if fileManager.fileExists(atPath: filePath) {
            _ = try fileManager.replaceItemAt(destinationURL, withItemAt: source, backupItemName: nil, options: .usingNewMetadataOnly)
        } else {
            try fileManager.moveItem(at: source, to: destinationURL)
        }
    }
    
    /// Creates a directory at the specified path if it does not already exist.
    /// Intermediate directories are created automatically as needed.
    /// - Parameter folderName: The full absolute path where the directory should be created.
    @objc public func createFolderIfDoesNotExist(_ folderName: String) {
        guard !folderName.isEmpty else { return }
        let fileManager = FileManager.default
        if !fileManager.fileExists(atPath: folderName) {
            try? fileManager.createDirectory(atPath: folderName, withIntermediateDirectories: true, attributes: nil)
        }
    }
    
    /// Deletes a specific file from the application's internal storage sandbox.
    /// - Parameters:
    ///   - fileName: The name of the file to delete (including extension).
    ///   - folder: The subfolder namespace within the workspace where the file is located.
    /// - Throws: An `NSError` if the file manager fails to remove the item.
    @objc public func deleteFile(_ fileName: String, inFolder folder: String) throws {
        let path = fullPathInStorageForFilePath(fileName, inFolder: folder)
        try FileManager.default.removeItem(atPath: path)
    }
}
